text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+users can perform (or prepare to perform) nonblocking operations during
+destructors when they are called in an async context.
+
+It's important to note, however, that `poll_drop_ready` may be called even
+after it has returned `Poll::Ready(())`. This is different from the `drop`
+method, which is generally guaranteed to be called only once. Users
+implementing `poll_drop_ready` should take care to ensure that it has "fuse"
+semantics - that once it returns `Ready(())`, it continues to return that value
+thereafter.
+
+These additional APIs are also added to the `std::mem` module:
+
+```rust
+// an empty async function
+async fn drop_async<T>(to_drop: T) { }
+
+fn poll_drop_ready<T>(&mut self, cx: &mut Context<'_>) -> Poll<()>
+    where T: ?Sized
+{
+    // implemented through a lang item
+}
+```
+
+The `drop_async` function is analogous to the `drop` function, and is also
+added to the prelude as well. This function drops the value in an async
+context, guaranteeing that its `poll_drop_ready` method will be called.
+
+The `poll_drop_ready` function calls this value's "drop ready glue" - it calls
+`poll_drop_ready` on this value and also on all of its fields, recursively. The
+exact ordering and semantics of this glue are specified in the reference
+section.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Guarantees about when `poll_drop_ready` is called
+
+In general, users cannot assume that `poll_drop_ready` will ever be called,
+just as they cannot assume that destructors will run at all. However, they can
+assume less about `poll_drop_ready` than they can about `drop`, and that should
+be noted.
+
+In particular, users cannot safely assume that their values will be dropped in
+an async context, regardless of how they are intended to be used. It's this
+easy to supress the `poll_drop_ready` call of a value:
+
+```rust
+async fn ignore_poll_drop_ready<T>(x: T) {
+    // by passing `x` to `drop`, a non-async context, `poll_drop_ready` is
+    // never called when dropping `x`.
+    drop(x);
+}
+```
+
+However, we do guarantee that values dropped in an async context *will* have
+`poll_drop_ready` called, and we even guarantee the drop order between
+variables and between fields of a type. When the user sees a value go out of
+scope *in an async context*, they know that `poll_drop_ready` is called. And we
+do guarantee that the destructors of all fields of that type will be called.
+
+## Changes to destructor calls in async contexts
+
+In async contexts, we change the generated destructor calls to insert, before
+each call, this code:
+
+```rust
+while let Poll::Pending = mem::poll_drop_ready(obj, cx) {
+    yield Poll::Pending;
+}
+```
+
+This "prepares" the object to be destroyed by calling its destructor, calling
+`poll_drop_ready` on the object and all of its fields. This is called before
+any destructors are run on that object.
```

Comment:
Would the reference counted pointers have a forwarding implementation of `poll_drop_ready` to the inner type? Is it possible to provide such implementation?

The proposed drop order makes it challenging if not impossible. `Drop::poll_drop_ready` on the fields happens before `Drop::drop` of the struct, so on the one hand, the async drop process of the inner type might have already started, on the other hand, the shared pointer themselves must still remain fully functional.

Reply:
* The `Arc<T>` recovers unique ownership over `T` inside `Arc<T>::drop` implementation (when it decrements counter to zero).
* Forwarding the `Arc<T>::poll_drop_ready` to `T::poll_drop_ready` requires a unique access, but at this point ownership is still shared.

Additionally decrementing the counter in `Arc<T>::poll_drop_ready` is impossible, because `Arc<T>` is accessible and must remain valid (according to the proposed drop order).

Reply:
This may be a reason we should return from `poll_drop_ready` to `poll_drop`, implemented by default by forwarding to the normal drop.

Rc/Arc would need to be changed to this basic algorithm:

```rust
if strong_count > 0:
    decrement strong_count

if strong_count == 0:
    ready!(inner.poll_drop)
    deallocate
```

This may be possible to implement correctly with poll_drop_ready, too.